server/watch_volfile_changes.py

```diff
@@ -0,0 +1,127 @@
+
+data = {}
+info_file_path = os.path.join(VOLINFO_DIR, "%s.info" % volname)
+
+def send_sighup():
+    """ Send SIGHUP to Brick & SHD Process notifing change in volfiles """
+    brick_process = "glusterfsd"
+    shd_process = "glusterfs"
+    brick_pid = ""
+    shd_pid = ""
+
+    try:
+        out, err, pid = execute("pidof", brick_process)
+        if out:
+            logging.info(logf("Brick PID", brick_pid=out))
+            brick_pid = out
+    except CommandException as err:
+        logging.error(logf(
+            "Failed to get brick process PID",
+            error=err
+        ))
+
+    try:
+        out, err, pid = execute("pidof", shd_process)
+        if out:
+            logging.info(logf("SHD PID", shd_pid=out))
+            shd_pid = out
+    except CommandException as err:
+        logging.error(logf(
+            "Failed to get SHD process PID",
+            error=err
+        ))
+
+    if brick_pid:
+        try:
+            execute("kill", "-HUP", str(brick_pid))
+            logging.info(logf("Send SIGHUP to brick process(GlusterFSD)"))
+        except CommandException as err:
+            logging.error(logf(
+                "Failed to send SIGHUP to brick process",
+                error=err
+            ))
+
+    if shd_pid:
+        try:
+            execute("kill", "-HUP", str(shd_pid))
+            logging.info(logf("Send SIGHUP to SHD process"))
+        except CommandException as err:
+            logging.error(logf(
+                "Failed to send SIGHUP to SHD process",
+                error=err
+            ))
+
+# TODO: Find more efficient ways to watch change in volfiles.
+# As mtime based watch will not work on Configmap,
+# since it is immutable & require pod refresh for mtime to be changed.
+def watch_volfile_changes():
+    """
+    Watch mounted configmap file for changes infinitely every 10 seconds.
+    If file is changed, regenerate volfiles & send SIGHUP
+    """
+
+    last_content = None
+
+    # Intialize last_content to avoid regenerating volfiles on first pass.
+    with open(info_file_path, "r") as info_file:
+        last_content = info_file.read()
+
+    while True:
+        with open(info_file_path, "r") as info_file:
+            content = info_file.read()
+
+            if last_content is not None and content != last_content:
+                last_content = content
+                logging.info(logf(
+                    "Detected change in configmap content. Regenerating volfiles."
+                ))
+
+                with open(info_file_path) as info_file:
+                    data = json.load(info_file)
+
+                create_brick_volfile(storage_unit_volfile_path, volname, volume_id, brick_path, data)
+                create_shd_volfile(shd_volfile_path, volname)
+                create_client_volfile(client_volfile_path, data)
+
+                # Send HUP to GlusterFSD & SHD processes
+                send_sighup()
+
+            else:
+                logging.info(logf(
```

Comment:
Make this debug. Otherwise, this will flood

Reply:
or remove

Reply:
Removed